namenode/namenode.go

```diff
@@ -44,27 +55,279 @@ func (server *NameNode) Create(path string, success *bool) error {
 
 func (server *NameNode) Exists(path string, exists *bool) error {
 	slog.Info("Exists request", "path", path)
-	server.globalRWLock.RLock()
-	defer server.globalRWLock.RUnlock()
+	server.inodeRWLock.RLock()
+	defer server.inodeRWLock.RUnlock()
 	_, *exists = server.inodes[path]
 	return nil
 }
 
+// TODO: 2pc, notify all the datanodes to delete the blocks before responding
 // This RPC doesn't return anything
 func (server *NameNode) Delete(path string, unused *bool) error {
 	slog.Info("Delete request", "path", path)
 
-	server.globalRWLock.Lock()
-	defer server.globalRWLock.Unlock()
+	server.inodeRWLock.Lock()
+	defer server.inodeRWLock.Unlock()
 
 	delete(server.inodes, path)
 	return nil
 }
 
 func (server *NameNode) GetAttributes(path string, fileAttributes *common.FileAttributes) error {
 	slog.Info("GetAttributes request", "path", path)
-	server.globalRWLock.RLock()
-	defer server.globalRWLock.RUnlock()
+	server.inodeRWLock.RLock()
+	defer server.inodeRWLock.RUnlock()
 	*fileAttributes = server.inodes[path].fileAttributes
 	return nil
 }
+
+func (server *NameNode) RegisterDataNode(dataNodeEndpoint string, success *bool) error {
+	slog.Info("RegisterDataNode request", "dataNodeEndpoint", dataNodeEndpoint)
+
+	server.inodeRWLock.Lock()
+	defer server.inodeRWLock.Unlock()
+
+	// TODO: Check if the datanode is already registered
+	server.datanodes[dataNodeEndpoint] = DataNodeInfo{
+		heartbeatReceived: true,
+		isAlive:           true,
+	}
+
+	*success = true
+	return nil
+}
+
+// NameNode receives block report from DataNode,
+// and updates all block storage info in the corresponding inode
+func (server *NameNode) ReportBlock(blockReport common.BlockReport, success *bool) error {
+	slog.Info("ReportBlock request", "dataNodeEndpoint", blockReport.Endpoint,
+		slog.Any("Block MetaData", blockReport.BlockMetadata))
```

Comment:
Why do we need `Any`?

Reply:
If you try to directly output `blockReport.BlockMetadata` here you'll see an error. slog cannot directly output some types, for example a map, and slog.Any() converts it to a string or something that we can output

Reply:
I see. Probably `slog` uses some serialization mechanism provided by go which can't serialize maps. Thanks for education.